feat(discover): project confirmation

wmak opened the pull request:
This adds a confirmation window when a user makes a discover query that spans > 10 projects > 30 days. The user can click `Nevermind` and change their query. or `do it` and the query will still be run.

![image](https://user-images.githubusercontent.com/4205004/84722099-477f7000-af50-11ea-85d8-7acdfd20d14a.png)

wmak committed:
wip(discover): Show a warning when a user tries expensive queries

- When the user has a query that spans >=10 projects over 30 days show
  them a warning asking them to confirm

wmak committed:
cleanup

wmak committed:
wip: Use an await timeout so selection doesn't query

- What was happening before was that:
  - after confirming a long query `confirmQuery` woulld be left as true.
  - when changing state `canLoadEvents` would happen after the event
    values propagate to the table
  - but since confirmQuery was still True, and the query has changed, an
    api call would be made before the modal opens
  - Using this timeout so that confirmQuery will get set to False for
    when the query changes again, for queries that don't get the modal
    `confirmQuery` should get set to True very quickly still

wmak committed:
cleanup

billyvg commented:
@wmak Can you use the `setState` callback here instead?

billyvg commented on src/sentry/static/sentry/app/utils/projects.tsx:
```diff
@@ -401,7 +401,7 @@ type FetchProjectsOptions = {
 } & Pick<Props, 'limit' | 'allProjects'>;
 
-async function fetchProjects(
+export async function fetchProjects(
```
This is not intended to be exposed, can you explain a bit about what you're trying to do with projects?  Fetching all projects is not ideal as it will be painful for large orgs.

wmak replied:
Hrm I need to be able to know the counts of projects depending on the counts (but only so far as whether its >10 or not). so when a query is for all projects, whether that's 15 projects, or just 5 for an org.

billyvg replied:
😬 would it be overkill to add an API that returns 2 values, # of projects in org and # of projects you're a member of? Otherwise we could be transferring/serializing thousands of projects for just the length.

wmak replied:
Thinking about it more, that's the better solution I think :+1: I'll flip the use of fetchProjects over to a different function/endpoint

markstory commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -91,13 +100,75 @@ class Results extends React.Component<Props, State> {
+    const currentQuery = eventView.getEventsAPIPayload(location);
+
+    const duration =
+      currentQuery.statsPeriod && currentQuery.statsPeriod.endsWith('d')
```
You need to use `decodeScalar()` to convert the string | string[] into a string

markstory commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -91,13 +100,75 @@ class Results extends React.Component<Props, State> {
+
+    // Any time confirmed Query is enabled, need to flip it off after the fetch starts
+    await doTimeout(0);
+    this.setState({confirmedQuery: false});
```
Could you toggle the state before `fetchProjects` so that after you know the project length you can settle the state without an additional wait?

wmak replied:
Hmm not sure if I misunderstood, wouldn't toggling the state pre `fetchProjects` result in querying without confirmation?

wmak committed:
ref: Adding confirmed to the other api loading components

- Using setState callback
- TODO: tests
- TODO: projectCount endpoint

wmak committed:
feat: Adding an option to the projects endpoint for counts

- I could introduce this as a new endpoint but figured putting it
  besides the existing project endpoint would result in the least amount
  of code repeat

wmak committed:
ref: Removing extra print statement

wmak commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -298,10 +367,29 @@ class Results extends React.Component<Props, State> {
+              onCancel={this.longQueryCancelled}
+              message={
+                <p>
+                  You've created a query that will search for events made{' '}
+                  <strong>over more than 30 days</strong> for{' '}
```
Is there a better way to do this? This was recommended by `prettier` ¯\\_(ツ)_/¯

markstory replied:
Not really. But you should use tct() for this so translations can be done.

wmak committed:
ref: Adding some test

wmak committed:
ref: Adding a test for the new count functionality

markstory commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -298,10 +366,29 @@ class Results extends React.Component<Props, State> {
             </ContentBox>
+            <Confirm
+              priority="primary"
+              header={<strong>May lead to thumb twiddling</strong>}
```
```suggestion
              header={<strong>{t('May lead to thumb twiddling')}</strong>}
```

wmak replied:
:+1: updated, also added it to `confirmText` and `cancelText`

wmak committed:
ref: Moving projects count to a new endpoint, adding tct/t

wmak committed:
Merge branch 'master' into feat/project-confirmation

billyvg commented on src/sentry/static/sentry/app/utils/projects.tsx:
```diff
@@ -411,6 +415,7 @@ async function fetchProjects(
     cursor?: typeof cursor;
     per_page?: number;
     all_projects?: number;
+    get_counts?: number;
```
Is this used?

wmak replied:
Missed this during an earlier change, :+1: removed

billyvg commented on src/sentry/static/sentry/app/utils/projects.tsx:
```diff
@@ -394,6 +394,10 @@ class Projects extends React.Component<Props, State> {
 
 export default withProjects(withApi(Projects));
 
+export async function fetchProjectsCount(api: Client, orgId: string) {
+  return await api.requestPromise(`/organizations/${orgId}/projects-count/`);
+}
```
Can we move this to `app/actionCreators/projects`? Would like to keep this file only related to `<Projects>`

billyvg commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -299,10 +368,35 @@ class Results extends React.Component<Props, State> {
+              header={<strong>{t('May lead to thumb twiddling')}</strong>}
+              confirmText={t('Do it')}
+              cancelText={t('Nevermind')}
+              onConfirm={this.longQueryConfirmed}
+              onCancel={this.longQueryCancelled}
```
nb/nit: We generally name this with a `handle-` prefix

billyvg commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -299,10 +368,35 @@ class Results extends React.Component<Props, State> {
+              onCancel={this.longQueryCancelled}
+              message={
+                <p>
+                  {tct(
+                    `You've created a query that will search for events made
+                    [dayLimit] for [projectLimit].
+                    A lot has happened during that time, so this might take awhile.
+                    Are you sure you want to do this?`,
+                    {
+                      dayLimit: <strong>{t('over more than 30 days')}</strong>,
+                      projectLimit: <strong>{t('more than 10 projects')}</strong>,
+                    }
+                  )}
```
This can be simplified to 
```suggestion
                  {tct(
                    `You've created a query that will search for events made
                    [dayLimit:over more than 30 days] for [projectLimit:more than 10 projects].
                    A lot has happened during that time, so this might take awhile.
                    Are you sure you want to do this?`,
                    {
                      dayLimit: <strong/>,
                      projectLimit: <strong/>,
                    }
                  )}
```

This way, when we extract the string for translators, they see the entire sentence, instead of breaking out `dayLimit` and `projectLimit` into separate strings.

billyvg reviewed (commented):
Mostly style related comments, otherwise lgtm

markstory commented on tests/sentry/api/endpoints/test_organization_projects.py:
```diff
@@ -182,6 +182,20 @@ def test_user_projects(self):
+class OrganizationProjectsCountTest(APITestCase):
+    @fixture
+    def org(self):
+        return self.create_organization(owner=self.user, name="baz")
```
The base tests cases have an existing factory property for `self.organization` and `self.team` do you need to redefine them?

wmak replied:
Oh, good catch updated `OrganizationProjectsTest` too :+1:

wmak committed:
ref: Addressing PR comments

- unused var in fetchProjcets
- moving fetchProjectCounts to actionCreators
- renaming functions to handled to match convention
- fixing tct to correct usage
- extra fixture functions in tests

billyvg commented on src/sentry/static/sentry/app/components/charts/eventsRequest.tsx:
```diff
@@ -236,9 +242,13 @@ class EventsRequest extends React.PureComponent<EventsRequestProps, EventsReques
+    const {api, confirmedQuery, ...props} = this.props;
     let timeseriesData: EventsStats | MultiSeriesEventsStats | null = null;
 
+    if (confirmedQuery === false) {
```
Can we add a comment here? (Or maybe in the type def) I think we should clarify that we do a strict equality check for `false` because the default behavior is to not wait for confirmation (e.g. when it's undefined).

billyvg reviewed (approved):
This looks good to me -- it'd be great to have a follow-up PR with some tests

wmak committed:
Merge branch 'master' into feat/project-confirmation

wmak committed:
ref: Adding more comments

wmak committed:
fix: Setting a function to avoid updating during state transition

- Sets the open function in the Results component to be used in
  canLoadEvents to resolve a warning about not being able to update
  during an existing state transition

wmak committed:
ref: Adding tests

k-fish commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -99,9 +106,70 @@ class Results extends React.Component<Props, State> {
   }
 
-  async fetchTotalCount() {
+  canLoadEvents = async () => {
```
`canLoadEvents` sounds like is just verifying whether events can be loaded, but it's actually fetching them? -> `loadEvents`?

k-fish commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -99,9 +106,70 @@ class Results extends React.Component<Props, State> {
+      this.setState({confirmedQuery: false});
+    });
+    if (needConfirmation) {
+      this.openConfirm();
```
This should be fine so long as `<Confirm>` is rendered and calls `setOpenFunction` before this function, just double check that `openConfirm` is set when the component calls `componentDidMount` as that should be the first lifecycle hook which calls this function, as far as my rudimentary React knowledge goes.

wmak replied:
:+1: confirmed, added a debugger in `componentDidMount` and checked that `openConfirm` was already set

k-fish commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -99,9 +106,70 @@ class Results extends React.Component<Props, State> {
+    this.setState({needConfirmation: false, confirmedQuery: true}, () => {
+      this.setState({confirmedQuery: false});
+    });
+    this.fetchTotalCount(true);
```
Should you move the check for a delta in `confirmedQuery` into the `componentDidUpdate` conditional to be consistent with the other views instead of calling fetch directly here? Also, if not, you may want to add a comment explaining why `confirmedQuery` is being set to `true` then back to `false` in the callback here as well (to explain that this is how clicking `okay` on the confirm will actually run the query).

wmak replied:
Dang good call :+1: this is much cleaner, also means that `canLoadEvents` is also only checking that instead of fetching too :+1:

wmak committed:
ref: Move fetchTotalCount to update to match other components

markstory commented on src/sentry/static/sentry/app/utils/discover/eventView.tsx:
```diff
@@ -534,6 +534,24 @@ class EventView {
+  getDays(): number {
+    const statsPeriod = decodeScalar(this.statsPeriod);
+    const start = decodeScalar(this.start);
+    const end = decodeScalar(this.end);
```
Do you need to decodeScalar here? These properties should already be `string | undefined`

markstory commented on src/sentry/static/sentry/app/utils/discover/eventView.tsx:
```diff
@@ -534,6 +534,24 @@ class EventView {
+      return (
+        (new Date(end).getTime() - new Date(start).getTime()) / (24 * 60 * 60 * 1000)
+      );
+    } else {
+      return 0;
+    }
```
```suggestion
    }
    return 0;
```

nit: You don't need an else as all other branches returned

markstory commented on src/sentry/static/sentry/app/views/eventsV2/results.tsx:
```diff
@@ -97,12 +104,69 @@ class Results extends React.Component<Props, State> {
+        projectLength === 0 ||
+        (projectLength === 1 && currentQuery.project[0] === '-1')
+      ) {
+        const results = await fetchProjectsCount(api, organization.slug);
```
Should you catch this request failing? Right now the error will bubble up to the page level error boundary. You could catch it locally here and assume the next query is safe.

wmak committed:
Merge branch 'master' into feat/project-confirmation

wmak committed:
ref: Removing decodeScalar, and adding a try/catch for project-count

wmak committed:
Merge branch 'master' into feat/project-confirmation

wmak committed:
Merge branch 'master' into feat/project-confirmation
